includes/api/class-wc-admin-rest-onboarding-levels-controller.php

```diff
@@ -0,0 +1,256 @@
+<?php
+/**
+ * REST API Onboarding Levels Controller
+ *
+ * Handles requests to /onboarding/levels
+ *
+ * @package WooCommerce Admin/API
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Onboarding Levels controller.
+ *
+ * @package WooCommerce Admin/API
+ * @extends WC_REST_Data_Controller
+ */
+class WC_Admin_REST_Onboarding_Levels_Controller extends WC_REST_Data_Controller {
+	/**
+	 * Endpoint namespace.
+	 *
+	 * @var string
+	 */
+	protected $namespace = 'wc/v4';
+
+	/**
+	 * Route base.
+	 *
+	 * @var string
+	 */
+	protected $rest_base = 'onboarding/levels';
+
+	/**
+	 * Register routes.
+	 *
+	 * @since 3.5.0
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Get an array of all levels and child tasks.
+	 *
+	 * @todo Status values below should pull from the database task status once implemented.
+	 */
+	public function get_levels() {
+		$levels = array(
+			array(
+				'id'    => 'account',
+				'tasks' => array(
+					array(
+						'id'           => 'create_account',
+						'label'        => __( 'Create an account', 'woocommerce-admin' ),
+						'description'  => __( 'Speed up & secure your store', 'woocommerce-admin' ),
+						'illustration' => '',
+						'status'       => 'visible',
+						'is_required'  => false,
+					),
+				),
+			),
+			array(
+				'id'    => 'storefront',
+				'tasks' => array(
+					array(
+						'id'           => 'add_products',
+						'label'        => __( 'Add your products', 'woocommerce-admin' ),
+						'description'  => __( 'Bring your store to life', 'woocommerce-admin' ),
+						'illustration' => '',
+						'status'       => 'visible',
+						'is_required'  => true,
+					),
+					array(
+						'id'           => 'customize_appearance',
+						'label'        => __( 'Customize Appearance', 'woocommerce-admin' ),
+						'description'  => __( 'Ensure your store is on-brand', 'woocommerce-admin' ),
+						'illustration' => '',
+						'status'       => 'visible',
+						'is_required'  => false,
+					),
+				),
+			),
+			array(
+				'id'    => 'checkout',
+				'tasks' => array(
```

Comment:
What do you think about adding an additional filter for each of these subtask arrays?

`woocommerce_onboarding_checkout_tasks` or something? That way an extension can hook into a specific section for example.

Reply:
I was thinking it would be trivial to find and modify array items from `woocommerce_onboarding_levels`, but maybe this would be easier if we make this an associate array?

```php
$levels = array(
    ...
    'checkout' => array(...)
);
```

Then we can modify in the filter accordingly:

```php
add_filter( 'woocommerce_onboarding_levels', function( $levels ) {
    $levels[ 'checkout' ][] = array(...);
    $levels[ 'boss-level' ] = array(...);
    return $levels
} );
```

Might also need to convert this back to a non-associative array when preparing to retain order in JSON, but I can test if you like this method.

Reply:
An associate array works for me. I think it saves a little time / is a bit clearer which level you are modifying, so devs don't need to pull out the correct index key first.

Reply:
Added this in https://github.com/woocommerce/woocommerce-admin/pull/1938/commits/97c6969c0ea976d10d71e130b4dedd623b08a04b.

Reply:
@joshuatf The response seems not quite right (and the PHP tests are failing:

See account vs checkout

<img width="841" alt="Screen Shot 2019-04-03 at 8 12 22 AM" src="https://user-images.githubusercontent.com/689165/55477956-65c9ac00-55e8-11e9-8df5-f32483cd0a65.png">

<img width="539" alt="Screen Shot 2019-04-03 at 8 13 16 AM" src="https://user-images.githubusercontent.com/689165/55477979-724e0480-55e8-11e9-873f-8e1c09a76114.png">

Reply:
I think everything is in the response - was the ID out of order the issue?  We'll lose this ordering once we're in JSON, but went ahead and forced ID to the beginning for readability prior to that.

Also fixed tests for adding in levels to the associative array.

Reply:
Wow - My eyes totally skipped over `"id": "account"` there so I thought it was missing. Yeah that's what I meant.